## Flight Operations Data Analysis

This notebook will guide you through a complete data analysis workflow on the provided Flight Operations dataset using Pandas. We will cover loading, inspecting, cleaning, transforming, and analyzing the data to identify patterns and key findings.

### 1. Load and Inspect the Dataset

In [1]:
import pandas as pd

# Load the dataset
file_path = 'Day10_Flight_Operations_Dataset.csv'
try:
    df = pd.read_csv(file_path)
    print(f"Dataset '{file_path}' loaded successfully.")
except FileNotFoundError:
    print(f"Error: The file '{file_path}' was not found. Please ensure it's uploaded correctly.")
    df = None

Dataset 'Day10_Flight_Operations_Dataset.csv' loaded successfully.


In [2]:
if df is not None:
    # Display the first 5 rows of the dataset
    print("\nFirst 5 rows of the dataset:")
    display(df.head())

    # Display basic information about the dataset (columns, non-null counts, dtypes)
    print("\nDataset Information:")
    df.info()

    # Display descriptive statistics
    print("\nDescriptive Statistics:")
    display(df.describe(include='all'))


First 5 rows of the dataset:


,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5



Dataset Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180 entries, 0 to 179
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Flight_ID               180 non-null    object 
 1   Flight_Date             180 non-null    object 
 2   Airline                 180 non-null    object 
 3   Origin                  180 non-null    object 
 4   Destination             180 non-null    object 
 5   Aircraft                180 non-null    object 
 6   Travel_Class            180 non-null    object 
 7   Passengers              180 non-null    int64  
 8   Seat_Capacity           180 non-null    int64  
 9   Average_Ticket_Price    180 non-null    int64  
 10  Delay_Minutes           173 non-null    float64
 11  Flight_Status           180 non-null    object 
 12  Weather                 180 non-null    object 
 13  Booking_Channel         180 non-null    object 
 14  Avg_Baggage_Kg      

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
count,180,180,180,180,180,180,180,180.000000,180.000000,180.000000,173.000000,180,180,180,180.000000,180,180.000000
unique,180,111,6,11,8,4,3,NaN,NaN,NaN,NaN,3,5,4,NaN,4,NaN
top,FL0001,2026-02-10,Vistara,Delhi,Delhi,Airbus A319,Economy,NaN,NaN,NaN,NaN,On Time,Rain,Online Travel Portal,NaN,Vegetarian,NaN
freq,1,6,37,43,51,60,134,NaN,NaN,NaN,NaN,114,43,50,NaN,52,NaN
mean,NaN,NaN,NaN,NaN,NaN,NaN,NaN,122.772222,197.238889,6111.677778,19.335260,NaN,NaN,NaN,15.150000,NaN,3.494444
std,NaN,NaN,NaN,NaN,NaN,NaN,NaN,45.635651,15.408176,3615.220040,29.029712,NaN,NaN,NaN,5.705506,NaN,1.169908
min,NaN,NaN,NaN,NaN,NaN,NaN,NaN,45.000000,180.000000,1949.000000,0.000000,NaN,NaN,NaN,5.000000,NaN,2.000000
25%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,78.750000,186.000000,3815.500000,0.000000,NaN,NaN,NaN,10.000000,NaN,2.000000
50%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,125.000000,189.000000,5380.000000,5.000000,NaN,NaN,NaN,15.500000,NaN,3.000000
75%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,162.250000,210.000000,6790.000000,20.000000,NaN,NaN,NaN,20.000000,NaN,5.000000


### 3. Data Transformation and Feature Engineering

In [4]:
if df is not None:
    print("\n--- Data Transformation ---")

    # 1. Create 'Occupancy_Rate' feature
    df['Occupancy_Rate'] = (df['Passengers'] / df['Seat_Capacity']) * 100
    print("Created 'Occupancy_Rate' column.")

    # 2. Extract 'Day_of_Week' from 'Flight_Date'
    df['Day_of_Week'] = df['Flight_Date'].dt.day_name()
    df['Day_of_Week'] = df['Day_of_Week'].astype('category') # Convert to category for efficiency
    print("Created 'Day_of_Week' column.")

    # Display the first few rows with new features
    print("\nFirst 5 rows with new features:")
    display(df.head())


--- Data Transformation ---
Created 'Occupancy_Rate' column.
Created 'Day_of_Week' column.

First 5 rows with new features:


,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction,Occupancy_Rate,Day_of_Week
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3,116.666667,Monday
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5,33.333333,Sunday
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2,30.454545,Tuesday
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2,69.545455,Saturday
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5,98.888889,Friday


### 4. Exploratory Data Analysis (EDA) - Overall Metrics

In [5]:
if df is not None:
    print("\n--- Overall Flight Operations Metrics ---")

    # Calculate overall average delay
    avg_delay = df['Delay_Minutes'].mean()
    print(f"Average Flight Delay: {avg_delay:.2f} minutes")

    # Calculate overall average ticket price
    avg_ticket_price = df['Average_Ticket_Price'].mean()
    print(f"Average Ticket Price: ${avg_ticket_price:.2f}")

    # Calculate overall average occupancy rate
    avg_occupancy_rate = df['Occupancy_Rate'].mean()
    print(f"Average Occupancy Rate: {avg_occupancy_rate:.2f}%")

    # Calculate overall average passenger satisfaction
    avg_satisfaction = df['Passenger_Satisfaction'].mean()
    print(f"Average Passenger Satisfaction (1-5 scale): {avg_satisfaction:.2f}")

    # Calculate the percentage of delayed flights
    total_flights = df.shape[0]
    delayed_flights = df[df['Flight_Status'] == 'Delayed'].shape[0]
    percentage_delayed = (delayed_flights / total_flights) * 100
    print(f"Percentage of Delayed Flights: {percentage_delayed:.2f}%")


--- Overall Flight Operations Metrics ---
Average Flight Delay: 18.78 minutes
Average Ticket Price: $6111.68
Average Occupancy Rate: 62.66%
Average Passenger Satisfaction (1-5 scale): 3.49
Percentage of Delayed Flights: 32.78%


### 5. Analyzing Trends by Categories (Grouping and Aggregation)

#### 5.1. Performance by Airline

In [6]:
if df is not None:
    print("\n--- Airline Performance Analysis ---")
    airline_performance = df.groupby('Airline').agg(
        Average_Delay=('Delay_Minutes', 'mean'),
        Average_Ticket_Price=('Average_Ticket_Price', 'mean'),
        Average_Passenger_Satisfaction=('Passenger_Satisfaction', 'mean'),
        Total_Flights=('Flight_ID', 'count')
    ).sort_values(by='Total_Flights', ascending=False)

    print("\nAirline Performance (sorted by Total Flights):")
    display(airline_performance)


--- Airline Performance Analysis ---

Airline Performance (sorted by Total Flights):


/tmp/ipykernel_829/2865957064.py:3: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  airline_performance = df.groupby('Airline').agg(


,Average_Delay,Average_Ticket_Price,Average_Passenger_Satisfaction,Total_Flights
Airline,,,,
Vistara,17.891892,6265.648649,3.378378,37
IndiGo,25.031250,5931.656250,3.343750,32
SpiceJet,22.343750,6189.750000,3.531250,32
Akasa Air,14.096774,5841.000000,3.419355,31
Air India Express,11.291667,7388.458333,3.666667,24
Air India,20.583333,5083.083333,3.750000,24


#### 5.3. Impact of Weather Conditions

In [8]:
if df is not None:
    print("\n--- Weather Impact Analysis ---")
    weather_impact = df.groupby('Weather').agg(
        Average_Delay=('Delay_Minutes', 'mean'),
        Average_Satisfaction=('Passenger_Satisfaction', 'mean'),
        Total_Flights=('Flight_ID', 'count')
    ).sort_values(by='Average_Delay', ascending=False)

    print("\nAnalysis by Weather Condition:")
    display(weather_impact)


--- Weather Impact Analysis ---

Analysis by Weather Condition:


/tmp/ipykernel_829/990776495.py:3: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  weather_impact = df.groupby('Weather').agg(


,Average_Delay,Average_Satisfaction,Total_Flights
Weather,,,
Rain,21.418605,3.372093,43
Storm,19.600000,3.166667,30
Cloudy,18.842105,3.842105,38
Clear,18.029412,3.352941,34
Fog,15.485714,3.685714,35


### 6. Key Findings and Observations

Based on the analysis, here are some key observations:

1.  **Overall Operations**: The average flight delay across all operations is **18.78 minutes**, with **32.78%** of all flights experiencing a delay, indicating that delays are a significant operational challenge. The average passenger satisfaction is moderate at **3.49 out of 5**, and the average occupancy rate stands at **62.66%**.

2.  **Impact of Flight Status on Satisfaction**: Unsurprisingly, **Delayed flights have the lowest average passenger satisfaction (3.27)** compared to On-Time flights (3.59) and even Cancelled flights (3.86). This suggests that effective delay management and communication are crucial for passenger experience. Interestingly, cancelled flights have a higher satisfaction, possibly due to prompt refunds or alternative arrangements.

3.  **Travel Class Dynamics**: **Business Class has significantly higher average ticket prices ($16,780)** than Premium Economy ($8,736) and Economy ($4,550). While Business Class has a similar occupancy rate to Premium Economy (around 67%), its passenger satisfaction (3.27) is notably lower than Premium Economy (3.69) and Economy (3.46). This indicates that higher-paying Business Class passengers may have higher expectations that are not always met.

4.  **Airline Performance Variation**: Among airlines, **Air India Express and Air India exhibit the highest average passenger satisfaction (3.67 and 3.75 respectively)**, despite Air India having a relatively higher average delay (20.58 minutes). IndiGo, while having a high number of flights, shows a higher average delay (25.03 minutes) and slightly lower satisfaction (3.34). Vistara has the most flights but an average delay of 17.89 minutes and satisfaction of 3.38.

5.  **Weather's Role in Delays**: **Rainy weather leads to the highest average delays (21.42 minutes)**, closely followed by Stormy conditions (19.60 minutes). Cloudy and Clear weather conditions have similar average delays (around 18 minutes), while Fog results in the lowest average delay (15.49 minutes). Passenger satisfaction also tends to be lower during adverse weather, with Stormy weather having the lowest satisfaction (3.17).

6.  **Occupancy and Pricing**: While Economy class has the lowest average ticket price, it also has the lowest average occupancy rate (60.98%) compared to Business and Premium Economy (both around 67%). This suggests potential opportunities to optimize pricing or capacity in Economy class to boost occupancy.

7.  **Key Determinants of Satisfaction**: Delay minutes and travel class appear to be strong determinants of passenger satisfaction. Minimizing delays, especially during adverse weather, and addressing the specific expectations of Business Class passengers could significantly improve overall satisfaction.

#### 5.2. Impact of Flight Status and Travel Class

In [7]:
if df is not None:
    print("\n--- Flight Status Analysis ---")
    flight_status_analysis = df.groupby('Flight_Status').agg(
        Average_Delay=('Delay_Minutes', 'mean'),
        Average_Satisfaction=('Passenger_Satisfaction', 'mean'),
        Total_Flights=('Flight_ID', 'count')
    )
    print("\nAnalysis by Flight Status:")
    display(flight_status_analysis)

    print("\n--- Travel Class Analysis ---")
    travel_class_analysis = df.groupby('Travel_Class').agg(
        Average_Ticket_Price=('Average_Ticket_Price', 'mean'),
        Average_Occupancy_Rate=('Occupancy_Rate', 'mean'),
        Average_Satisfaction=('Passenger_Satisfaction', 'mean'),
        Total_Flights=('Flight_ID', 'count')
    ).sort_values(by='Average_Ticket_Price', ascending=False)

    print("\nAnalysis by Travel Class:")
    display(travel_class_analysis)


--- Flight Status Analysis ---

Analysis by Flight Status:


/tmp/ipykernel_829/1640586443.py:3: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  flight_status_analysis = df.groupby('Flight_Status').agg(


,Average_Delay,Average_Satisfaction,Total_Flights
Flight_Status,,,
Cancelled,5.000000,3.857143,7
Delayed,50.000000,3.271186,59
On Time,3.464912,3.587719,114



--- Travel Class Analysis ---

Analysis by Travel Class:


/tmp/ipykernel_829/1640586443.py:12: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  travel_class_analysis = df.groupby('Travel_Class').agg(


,Average_Ticket_Price,Average_Occupancy_Rate,Average_Satisfaction,Total_Flights
Travel_Class,,,,
Business,16780.000000,67.432522,3.272727,11
Premium Economy,8736.742857,67.625910,3.685714,35
Economy,4550.268657,60.976998,3.462687,134


### 2. Data Cleaning and Type Conversion

In [3]:
if df is not None:
    print("\n--- Data Cleaning and Type Conversion ---")

    # 1. Handle Missing Values in 'Delay_Minutes'
    # Impute missing 'Delay_Minutes' with the median value
    median_delay = df['Delay_Minutes'].median()
    df['Delay_Minutes'].fillna(median_delay, inplace=True)
    print(f"Missing values in 'Delay_Minutes' imputed with median: {median_delay}.")

    # 2. Convert 'Flight_Date' to datetime objects
    df['Flight_Date'] = pd.to_datetime(df['Flight_Date'])
    print("Converted 'Flight_Date' to datetime objects.")

    # 3. Convert object columns to 'category' type for memory efficiency and better analysis
    categorical_cols = [
        'Airline', 'Origin', 'Destination', 'Aircraft',
        'Travel_Class', 'Flight_Status', 'Weather',
        'Booking_Channel', 'Meal_Preference'
    ]
    for col in categorical_cols:
        df[col] = df[col].astype('category')
    print("Converted several object columns to 'category' type.")

    # Verify changes
    print("\nUpdated Dataset Information:")
    df.info()
    print("\nNumber of missing values after cleaning:")
    print(df.isnull().sum())


--- Data Cleaning and Type Conversion ---
Missing values in 'Delay_Minutes' imputed with median: 5.0.
Converted 'Flight_Date' to datetime objects.
Converted several object columns to 'category' type.

Updated Dataset Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180 entries, 0 to 179
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   Flight_ID               180 non-null    object        
 1   Flight_Date             180 non-null    datetime64[ns]
 2   Airline                 180 non-null    category      
 3   Origin                  180 non-null    category      
 4   Destination             180 non-null    category      
 5   Aircraft                180 non-null    category      
 6   Travel_Class            180 non-null    category      
 7   Passengers              180 non-null    int64         
 8   Seat_Capacity           180 non-null    int64         
 9  

/tmp/ipykernel_829/4081903710.py:7: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Delay_Minutes'].fillna(median_delay, inplace=True)
